### DISCUSSION TITLE
jamulus-headless command-line args

### DISCUSSION THREAD

**Discussion by sfrankel** [jamulussoftware/jamulus#3208] [General]
Hello -- what are the proper command-line args to run v3.10 jamulus-headless? My copy seems to be running, but it won't write to a log file, it won't create a recording directory or record, and I cannot connect to it from a Jamulus client. 

Here's the command I'm using
`% /usr/bin/jamulus-headless \
--welcomemessage "<h2>welcome to my server</h2>" \
--log /home/ubuntu/jamulus/logs/jam_20231211.txt \
--recording /home/ubuntu/jamulus/recordings/ \
--numchannels 16`

Shell output looks good
`- Starting in server mode by default (due to compile time option)
- welcome message: <h2>welcome to my server</h2>
- logging file name: /home/ubuntu/jamulus/logs/jam_20231211.txt
- recording directory name: /home/ubuntu/jamulus/recordings/
- maximum number of channels: 16
No GUI support compiled. Running in headless mode.
Using "xxx.xx.xx.xxx" as external IP.
Recording state: enabled
 *** Jamulus, Version 3.10.0
...
`

But it's not working for some reason; eg:
`% ll jamulus/logs/jam_20231211.txt 
-rw-rw-r-- 1 ubuntu ubuntu 0 Dec 12 01:55 jamulus/logs/jam_20231211.txt
`

This is on an AWS Lightsail instance running a freshly created OS-only Ubuntu 22.04 LTS. My user name on that instance is "ubuntu" and the logs/ and recordings/ directories I created are confirmed writable:
`drwxrwxr-x 4 ubuntu ubuntu 4096 Dec 11 23:46 jamulus/
...
drwxrwxr-x 2 ubuntu ubuntu 4096 Dec 11 23:46 recordings
drwxrwxr-x 2 ubuntu ubuntu 4096 Dec 12 01:55 logs`

Suggestions? TIA!
Scott

**Comment by mcfnord** [comment]
I believe the --log parameter does not work when running headlessly under systemctl. Instead, journalctl can be used.

I also believe a systemctl process cannot write to /home/ubuntu/ using the security provisions we specify by default. 

Enter this:

sudo systemctl edit --full jamulus-headless

There's a Home Directory protection in there that prevents writing to the home directory.

But why can't a client connect to it? I bet that's because you have not opened the port explicitly in aws. I don't use LightSail, but often create aws spot instances in EC2, but since I list them on a public directory, I don't need to explicitly open the port.

**Reply by rdica** [reply]
The logging parameter does in fact work and is independent of systemd journalling.

There's nothing in jamulus itself that prevents writing to a home directory, thats a systemd function of the service file configuration for jamulus. Based on the example given, you're not running the process from systemd but from a command line.

The log itself writes only client connections, not directory server registrations. If nothing can connect to your server no log entry will be written, hence 0 bytes in the log file. If no clients can connect, nothing can be recorded, so no directories/files will be created.

Your command line looks fine to me. Once you resolve your connection issues, you should begin to see client connections in log, and recordings dirs/files.

FYI, the systemd journal contains client connection data, directory server registration status (if using a directory server), and any other output the jamulus server itself may generate during the course of its execution.

**Reply by sfrankel** [reply]
Solved
Thanks for confirming the command line! Turns out I had to open a port range to allow external access

`Add a new IPv4 firewall rule to allow external user access
- AWS:Lightsail console -> Manage -> Networking -> Firewall
  - Click '+ Add rule'
  - Application: Custom
  - Protocol: UDP
  - Port or range: 22120-22130
  - Restrict to IP address: unchecked
  - Click Create`

**Reply by mcfnord** [reply]
Allowing that port range is also what I needed. 

Why doesn't 'range' appear on this page?:
https://jamulus.io/wiki/Unregistered-Servers

**Reply by rdica** [reply]
A range isn't required based on my testing.
Jamulus listens on a single port, opening the firewall to that single port is the minimum requirement when not using a directory server. Why @sfrankel chose to implement a port range can only be answered by him, but note that the range he used covers the default `udp/22124` port so it'd work. Specifying a single port would have worked as well.

Try it and let us know your results.

I submit there's no reason to mention a range on that page as there's no range requirement.

**Reply by mcfnord** [reply]
Ok, UDP with just 22124 open does work!

I may have mixed TCP into my tests. :(
